### ISSUE TITLE
Bug in the 0.24.0 and 0.25.0 providers?

### ISSUE THREAD

**Issue by ianchesal** [buildkite/terraform-provider-buildkite#371]
**Describe the bug**
I'm working with existing Terraform here and trying to upgrade from 0.23.0 to 0.24.0 or 0.25.0.

When I try to apply my changes using 0.24.0 and 0.25.0 I'm seeing two things:

1. An unexpectedly high amount of `(known after apply)` notices on parameters
2. An actual error from the provider on apply for the `.tags` attribute

**To Reproduce**

I'm using the following modular definition of a pipelne:

```
resource "buildkite_pipeline" "pipeline" {
  # See: https://registry.terraform.io/providers/buildkite/buildkite/latest/docs/resources/pipeline
  for_each = local.pipelines

  # Required settings
  default_branch = each.value.default_branch
  description    = each.value.description
  name           = each.key
  repository     = each.value.repository

  # Immutable settings
  steps = lookup(each.value, "custom_steps", lookup(each.value, "deflake_mode", false) ? file("./deflake-steps.yaml") : file("./default-steps.yaml"))

  # Optional settings
  allow_rebuilds             = lookup(each.value, "allow_rebuilds", true)
  branch_configuration       = lookup(each.value, "branch_configuration", "")
  default_timeout_in_minutes = lookup(each.value, "default_timeout_in_minutes", 0)
  maximum_timeout_in_minutes = lookup(each.value, "maximum_timeout_in_minutes", 0)
  tags                       = lookup(each.value, "tags", [])

  # When you push the same branch with two different HEADs, by default
  # Buildkite will start 2 builds. Buildkite lets you specify what should
  # happen to the first build: "skip" it (if the build hasn't started yet), or
  # cancel it (if the build did start). Most (all?) of the time, you don't care
  # about the difference between "skipping" and "cancelling" the first build,
  # you just want the first build to be interrupted one way or another. So we
  # provide a shortcut "only_build_latest_branch" which is by default on.
  skip_intermediate_builds                 = lookup(each.value, "only_build_latest_branch", lookup(each.value, "skip_intermediate_builds", true))
  skip_intermediate_builds_branch_filter   = lookup(each.value, "only_build_latest_branch_filter", lookup(each.value, "skip_intermediate_builds_branch_filter", ""))
  cancel_intermediate_builds               = lookup(each.value, "only_build_latest_branch", lookup(each.value, "cancel_intermediate_builds", true))
  cancel_intermediate_builds_branch_filter = lookup(each.value, "only_build_latest_branch_filter", lookup(each.value, "cancel_intermediate_builds_branch_filter", ""))

  provider_settings {
    build_branches                           = lookup(each.value, "build_branches", true)
    build_pull_requests                      = lookup(each.value, "build_pull_requests", true)
    build_tags                               = lookup(each.value, "build_tags", false)
    cancel_deleted_branch_builds             = lookup(each.value, "cancel_deleted_branch_builds", false)
    filter_condition                         = lookup(each.value, "filter_condition", "")
    filter_enabled                           = lookup(each.value, "filter_enabled", false)
    publish_commit_status                    = lookup(each.value, "publish_commit_status", true)
    publish_commit_status_per_step           = lookup(each.value, "publish_commit_status_per_step", false)
    pull_request_branch_filter_configuration = lookup(each.value, "pull_request_branch_filter_configuration", "")
    pull_request_branch_filter_enabled = lookup(each.value,
    "pull_request_branch_filter_enabled", false)
    # We'll default to true for this, because we are unable to think of a
    # scenario where you'd want multiple builds for the same commit running.
    # https://withpersona.slack.com/archives/C02GNMPGWFJ/p1684954136079899
    skip_pull_request_builds_for_existing_commits = lookup(each.value, "skip_pull_request_builds_for_existing_commits", true)
  }
}
```

And here's an example of a variables.tf entry to instantiate an instance of the module:

```
locals {
  pipelines = {
    "infrastructure-tools" = {
      repository     = "git@bitbucket.org:persona-id/persona-infrastructure-tools.git"
      description    = "The persona-infrastructure-tools repository pipelines"
      default_branch = "main"
    }
  }
}
```

When my provider is set to 0.23.0 a plan yields:

```
> tfp -target='buildkite_pipeline.pipeline["infrastructure-tools"]'
buildkite_pipeline.pipeline["infrastructure-tools"]: Refreshing state... [id=UGlwZWxpbmUtLS0wMTg5NDVjOC1jYzMyLTQyOTktYjExMi0xYTdjZjZlNmVhMDA=]

No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.
```

And if I upgrade to 0.24.0 I get:

```
> tfp -target='buildkite_pipeline.pipeline["infrastructure-tools"]' -out=tfplan
buildkite_pipeline.pipeline["infrastructure-tools"]: Refreshing state... [id=UGlwZWxpbmUtLS0wMTg5NDVjOC1jYzMyLTQyOTktYjExMi0xYTdjZjZlNmVhMDA=]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # buildkite_pipeline.pipeline["infrastructure-tools"] will be updated in-place
  ~ resource "buildkite_pipeline" "pipeline" {
        id                         = "UGlwZWxpbmUtLS0wMTg5NDVjOC1jYzMyLTQyOTktYjExMi0xYTdjZjZlNmVhMDA="
        name                       = "infrastructure-tools"
      ~ slug                       = "infrastructure-tools" -> (known after apply)
      + tags                       = []
        # (13 unchanged attributes hidden)

      ~ provider_settings {
          + build_pull_request_forks                      = (known after apply)
          + build_pull_request_labels_changed             = (known after apply)
          + build_pull_request_ready_for_review           = (known after apply)
          + cancel_deleted_branch_builds                  = false
          + prefix_pull_request_fork_branch_names         = (known after apply)
          + publish_blocked_as_pending                    = (known after apply)
          + separate_pull_request_statuses                = (known after apply)
          + skip_builds_for_existing_commits              = (known after apply)
          + trigger_mode                                  = (known after apply)
            # (8 unchanged attributes hidden)
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.
╷
│ Warning: Resource targeting is in effect
│
│ You are creating a plan with the -target option, which means that the result of this plan may not represent all of the changes requested by the current configuration.
│
│ The -target option is not for routine use, and is provided only for exceptional situations such as recovering from errors or mistakes, or when Terraform specifically suggests to
│ use it as part of an error message.
╵

─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Saved the plan to: tfplan

To perform exactly these actions, run the following command to apply:
    terraform apply "tfplan"
```

And trying to apply I get the following error:

```
> tfa tfplan && rm tfplan
buildkite_pipeline.pipeline["infrastructure-tools"]: Modifying... [id=UGlwZWxpbmUtLS0wMTg5NDVjOC1jYzMyLTQyOTktYjExMi0xYTdjZjZlNmVhMDA=]
╷
│ Warning: Applied changes may be incomplete
│
│ The plan was created with the -target option in effect, so some changes requested in the configuration may have been ignored and the output values may not be fully updated. Run
│ the following command to verify that no other changes are pending:
│     terraform plan
│
│ Note that the -target option is not suitable for routine use, and is provided only for exceptional situations such as recovering from errors or mistakes, or when Terraform
│ specifically suggests to use it as part of an error message.
╵
╷
│ Error: Provider produced inconsistent result after apply
│
│ When applying changes to buildkite_pipeline.pipeline["infrastructure-tools"], provider "provider[\"registry.terraform.io/buildkite/buildkite\"]" produced an unexpected new value:
│ .tags: was cty.SetValEmpty(cty.String), but now null.
│
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
╷
│ Error: Provider produced inconsistent result after apply
│
│ When applying changes to buildkite_pipeline.pipeline["infrastructure-tools"], provider "provider[\"registry.terraform.io/buildkite/buildkite\"]" produced an unexpected new value:
│ .provider_settings[0].cancel_deleted_branch_builds: was cty.False, but now null.
│
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
```

**Expected behavior**
I was expecting essentially no changes here on plan/apply as the module and variables haven't changed, only the provider version. And the docs for 0.24.0 don't make it look in any way different from 0.23.0 with respect to these parameters.

**Additional context**

```
> tf --version
Terraform v1.5.5
on darwin_arm64
+ provider registry.terraform.io/buildkite/buildkite v0.24.0
```

I have a valid API token in my environment.

**Comment by ianchesal** [comment]
And it looks like 0.25.0 came out 20 hours. Trying that!

**Comment by ianchesal** [comment]
Hmm. I still get a worryingly high list of in-place changes:

```
> tfp -target='buildkite_pipeline.pipeline["infrastructure-tools"]' -out=tfplan
buildkite_pipeline.pipeline["infrastructure-tools"]: Refreshing state... [id=UGlwZWxpbmUtLS0wMTg5NDVjOC1jYzMyLTQyOTktYjExMi0xYTdjZjZlNmVhMDA=]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # buildkite_pipeline.pipeline["infrastructure-tools"] will be updated in-place
  ~ resource "buildkite_pipeline" "pipeline" {
        id                         = "UGlwZWxpbmUtLS0wMTg5NDVjOC1jYzMyLTQyOTktYjExMi0xYTdjZjZlNmVhMDA="
        name                       = "infrastructure-tools"
      ~ slug                       = "infrastructure-tools" -> (known after apply)
        tags                       = []
        # (13 unchanged attributes hidden)

      ~ provider_settings {
          + build_pull_request_forks                      = (known after apply)
          + build_pull_request_labels_changed             = (known after apply)
          + build_pull_request_ready_for_review           = (known after apply)
          + cancel_deleted_branch_builds                  = false
          + prefix_pull_request_fork_branch_names         = (known after apply)
          + publish_blocked_as_pending                    = (known after apply)
          + separate_pull_request_statuses                = (known after apply)
          + skip_builds_for_existing_commits              = (known after apply)
          + trigger_mode                                  = (known after apply)
            # (8 unchanged attributes hidden)
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

And applying still reports that there's an error and an inconsistent result, so potentially still an issue in 0.25.0:

```
> tfa tfplan && rm tfplan
buildkite_pipeline.pipeline["infrastructure-tools"]: Modifying... [id=UGlwZWxpbmUtLS0wMTg5NDVjOC1jYzMyLTQyOTktYjExMi0xYTdjZjZlNmVhMDA=]
╷
│ Warning: Applied changes may be incomplete
│
│ The plan was created with the -target option in effect, so some changes requested in the configuration may have been ignored and the output values may not be fully updated. Run
│ the following command to verify that no other changes are pending:
│     terraform plan
│
│ Note that the -target option is not suitable for routine use, and is provided only for exceptional situations such as recovering from errors or mistakes, or when Terraform
│ specifically suggests to use it as part of an error message.
╵
╷
│ Error: Provider produced inconsistent result after apply
│
│ When applying changes to buildkite_pipeline.pipeline["infrastructure-tools"], provider "provider[\"registry.terraform.io/buildkite/buildkite\"]" produced an unexpected new value:
│ .provider_settings[0].cancel_deleted_branch_builds: was cty.False, but now null.
│
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.

```

**Comment by jradtilbrook** [comment]
Hi @ianchesal thanks for the thorough report here!

We found this and have released [v0.25.1](https://github.com/buildkite/terraform-provider-buildkite/releases/tag/v0.25.1) with fixes.

It also looks like you are hitting another path that is caused by not also performing a `terraform refresh` after upgrading the provider. Can you try this out with v0.25.1 and do a refresh prior to plan/apply and let us know if its resolves it.

(You'll still get a fair few diffs shown but they are computed values and shouldn't actually change anything - do another targeted plan/apply if you are worried and share it with us if it goes wrong)

**Comment by ianchesal** [comment]
Wow. Thanks for the quick turn around!

I've got a new issue with 0.25.1:

```
> tfp -target='buildkite_pipeline.pipeline["infrastructure-tools"]' -out=tfplan
buildkite_pipeline.pipeline["infrastructure-tools"]: Refreshing state... [id=UGlwZWxpbmUtLS0wMTg5NDVjOC1jYzMyLTQyOTktYjExMi0xYTdjZjZlNmVhMDA=]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # buildkite_pipeline.pipeline["infrastructure-tools"] will be updated in-place
  ~ resource "buildkite_pipeline" "pipeline" {
      ~ archive_on_delete          = true -> false
        id                         = "UGlwZWxpbmUtLS0wMTg5NDVjOC1jYzMyLTQyOTktYjExMi0xYTdjZjZlNmVhMDA="
        name                       = "infrastructure-tools"
      ~ slug                       = "infrastructure-tools" -> (known after apply)
        tags                       = []
        # (12 unchanged attributes hidden)

      ~ provider_settings {
          + build_pull_request_forks                      = (known after apply)
          + build_pull_request_labels_changed             = (known after apply)
          + build_pull_request_ready_for_review           = (known after apply)
          + cancel_deleted_branch_builds                  = false
          + prefix_pull_request_fork_branch_names         = (known after apply)
          + publish_blocked_as_pending                    = (known after apply)
          + separate_pull_request_statuses                = (known after apply)
          + skip_builds_for_existing_commits              = (known after apply)
          + trigger_mode                                  = (known after apply)
            # (8 unchanged attributes hidden)
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.
╷
│ Warning: Resource targeting is in effect
│
│ You are creating a plan with the -target option, which means that the result of this plan may not represent all of the changes requested by the current configuration.
│
│ The -target option is not for routine use, and is provided only for exceptional situations such as recovering from errors or mistakes, or when Terraform specifically suggests to
│ use it as part of an error message.
╵

─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Saved the plan to: tfplan

To perform exactly these actions, run the following command to apply:
    terraform apply "tfplan"
> tfa tfplan && rm tfplan
buildkite_pipeline.pipeline["infrastructure-tools"]: Modifying... [id=UGlwZWxpbmUtLS0wMTg5NDVjOC1jYzMyLTQyOTktYjExMi0xYTdjZjZlNmVhMDA=]
╷
│ Warning: Applied changes may be incomplete
│
│ The plan was created with the -target option in effect, so some changes requested in the configuration may have been ignored and the output values may not be fully updated. Run
│ the following command to verify that no other changes are pending:
│     terraform plan
│
│ Note that the -target option is not suitable for routine use, and is provided only for exceptional situations such as recovering from errors or mistakes, or when Terraform
│ specifically suggests to use it as part of an error message.
╵
╷
│ Error: Provider produced inconsistent result after apply
│
│ When applying changes to buildkite_pipeline.pipeline["infrastructure-tools"], provider "provider[\"registry.terraform.io/buildkite/buildkite\"]" produced an unexpected new value:
│ .provider_settings[0].cancel_deleted_branch_builds: was cty.False, but now null.
│
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
```

I did an `init -upgrade` and `refresh` before the plan/apply.

**Comment by jradtilbrook** [comment]
Thanks @ianchesal I can reproduce this so I'll look into a fix

**Comment by jradtilbrook** [comment]
Okay I see what is going on here now @ianchesal. You are using a bitbucket repo for this pipeline which the server has properly detected 👍 

However, not all available attributes under `provider_settings` apply when using a bitbucket repo. And `cancel_deleted_branch_builds` is one of those. You can see an example here in the UI:
<img width="847" alt="Screenshot 2023-08-25 at 11 07 28" src="https://github.com/buildkite/terraform-provider-buildkite/assets/6128798/e1894bec-4688-486a-8c35-c168becf34cd">

Through our upgrade to the newest terraform SDK package it enforces stricter validation on attributes. Specifically, if a user sets an attribute we cannot change that value in the state file. Because you have set `cancel_deleted_branch_builds = false`, we are trying to set it to null which is what the API returns because it doesn't apply to bitbucket.

We have plans to improve the `provider_settings` section greatly now we are using the new SDK version but haven't implemented it yet. We are still figuring out what it will look like - possibly using a completely different configuration schema.

Until then (hoping to nail that down in the next couple of weeks) I'm not sure if there is any bandaid we can apply here. Were you trying to upgrade for any features available in v0.25.1? Or  just trying to stay on top of your dependencies? That will help us prioritise things

**Comment by ianchesal** [comment]
> Were you trying to upgrade for any features available in v0.25.1? Or just trying to stay on top of your dependencies? That will help us prioritise things.

Simply trying to stay up-to-date on my dependencies. There was nothing specific in 0.24.0 that I was looking to use.

Thanks!
